xds/src/test/java/io/grpc/xds/XdsClientFederationTest.java

```diff
@@ -0,0 +1,153 @@
+import static org.mockito.Mockito.timeout;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.xds.XdsClient.ResourceWatcher;
+import io.grpc.xds.XdsListenerResource.LdsUpdate;
+import java.util.Collections;
+import java.util.Map;
+import java.util.UUID;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+/**
+ * Tests for xDS control plane federation scenarios.
+ */
+@RunWith(JUnit4.class)
+public class XdsClientFederationTest {
+
+  @Mock
+  private ResourceWatcher<LdsUpdate> mockDirectPathWatcher;
+
+  @Mock
+  private ResourceWatcher<LdsUpdate> mockWatcher;
+
+  private static final String SERVER_LISTENER_TEMPLATE_NO_REPLACEMENT =
+      "grpc/server?udpa.resource.listening_address=";
+
+  @Rule
+  public ControlPlaneRule trafficdirector = new ControlPlaneRule("test-server");
+
+  @Rule
+  public ControlPlaneRule directpathPa = new ControlPlaneRule(
+      "xdstp://server-one/envoy.config.listener.v3.Listener/test-server");
+
+  private XdsClient xdsClient;
+  private boolean originalFederationStatus;
+
+  @Before
+  public void setUp() throws XdsInitializationException {
+    MockitoAnnotations.initMocks(this);
+
+    originalFederationStatus = BootstrapperImpl.enableFederation;
+    BootstrapperImpl.enableFederation = true;
+
+    SharedXdsClientPoolProvider clientPoolProvider = new SharedXdsClientPoolProvider();
+    clientPoolProvider.setBootstrapOverride(defaultBootstrapOverride());
+    xdsClient = clientPoolProvider.getOrCreate().getObject();
+  }
+
+  @After
+  public void cleanUp() throws InterruptedException {
+    BootstrapperImpl.enableFederation = originalFederationStatus;
+  }
+
+  // Assures that resource deletions happening in one control plane do not trigger deletion events
+  /// in watchers of resources on other control planes.
+  @Test
+  public void isolatedResourceDeletions() throws InterruptedException {
+    // Add the mock watcher for the normal server resource. The test control plane will send
+    // a deletion event.
+    xdsClient.watchXdsResource(XdsListenerResource.getInstance(), "test-server", mockWatcher);
+    verify(mockWatcher, timeout(20000)).onResourceDoesNotExist("test-server");
+
+    // Add the watcher for the DirectPath server.
+    xdsClient.watchXdsResource(XdsListenerResource.getInstance(),
+        "xdstp://server-one/envoy.config.listener.v3.Listener/test-server", mockDirectPathWatcher);
+    verify(mockDirectPathWatcher, timeout(20000)).onResourceDoesNotExist(
+        "xdstp://server-one/envoy.config.listener.v3.Listener/test-server");
+    verify(mockWatcher, timeout(20000)).onResourceDoesNotExist("test-server");
```

Comment:
This does nothing, as it has already been verified and it isn't doing `times(2)` or such.